Growth estimators

adrian-lison opened the pull request:
## Description

This PR closes #768.

The PR adds a new option to calculate the growth rate in `estimate_infections()` and `simulate_infections()` that is based on an estimator by Parag et al. 2022.

## Initial submission checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have tested my changes locally (using `devtools::test()` and `devtools::check()`).
- [x] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required and rebuilt docs if yes (using `devtools::document()`).
- [x] I have followed the established coding standards (and checked using `lintr::lint_package()`).
- [x] I have added a news item linked to this PR.

## After the initial Pull Request 

- [ ] I have reviewed Checks for this PR and addressed any issues as far as I am able.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Added a selectable growth rate method (“infections” or “infectiousness”) based on Parag et al., available in options and simulations.
  - Introduced a numerically stable convolution utility (stable_convolve).
- Documentation
  - Expanded guidance for growth_method, updated defaults, and added references; clarified argument defaults elsewhere.
- Tests
  - Added comprehensive tests covering the new growth methods.
- Chores
  - Updated package authorship metadata.
  - Minor cleanup to reduce unnecessary dependency imports.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

adrian-lison commented:
@seabbs @sbfnk 

Just opened a draft PR while working on this. I have drafted the basic stan code, but couldn't test it locally yet because of issues with my rstan installation ^^

I have for now added an argument `growth_method` to the rt_opts(), because I didn't find it justified to create a completely new opts type. There are a couple of open questions to discuss, in particular how much of a problem it is that with the alternaitve approach the most recent growth rates are undefined due to the shift applied.

Will get back with more details soonish.

sbfnk commented:
Also please add a news item and add yourself as author to the `DESCRIPTION`.

seabbs commented on inst/stan/estimate_infections.stan:
```diff
@@ -258,26 +259,39 @@ generated quantities {
+        gt_rev_pmf_for_growth = gt_rev_pmf;
+      }
+  
+      // estimate growth from infections
```
to reduce noise here do you think we should wrap it in a calculate_growth wrapper?

seabbs reviewed (commented):
this is really nice sorry it got no response for ages. Some minor cleaning up suggestions. 

On an related but not for now point I wonder if we should be doing the prostprocessing in stan or R for these measures? I see upsides and downsides to both options.

adrian-lison commented on R/opts.R:
```diff
@@ -290,14 +304,16 @@ rt_opts <- function(prior = LogNormal(mean = 1, sd = 1),
-                    pop = 0) {
+                    pop = 0,
+                    growth_method = c("infections", "infectiousness")) {
   opts <- list(
     use_rt = use_rt,
     rw = rw,
     use_breakpoints = use_breakpoints,
     future = future,
     pop = pop,
-    gp_on = arg_match(gp_on)
+    gp_on = arg_match(gp_on),
+    growth_method = arg_match(growth_method)
```
I don't see a problem

adrian-lison commented:
@seabbs @sbfnk Thanks for your feedback and very sorry that this fell through the cracks! I have incorporated the suggested changes and made a couple of further edits:

- the shifting of the growth rate time series that results from the differencing is now internally handled by the `calculate_growth_` methods. So far, this was handled more ad-hoc by passing `seeding_time + 1` to `seeding_time`, which has potential for confusion.
- Since the `infectiousness` growth rate method has an additional shift by the mean generation time, causing the most recent growth rates to be `NaN`, there was another issue: if the generation time is sampled in the model, its mean may of course vary by more than 1 day, and therefore the number of undefined dates can differ across draws. This led to a bias towards shorter generation times in the growth rate summary, which only averaged over the draws with a defined rate for a given date. I addressed this for now by setting the growth rate to `NaN` consistently over all draws for these dates post-hoc (on the extracted samples).

Below is a comparison of the growth rate estimates from both methods on the `example_confirmed` cases. They are really similar except for the start of the time series (but this seems to match with the simulation from Parag et al. 2022).

<img width="2400" height="1200" alt="growth_rate_methods" src="https://github.com/user-attachments/assets/c6f705d7-d7d6-4a2e-9cb8-47193e0a96f1" />

sbfnk reviewed (approved):
Looks great, thank you very much!
